File: javascript/variable-length-quantity/variable-length-quantity.js
/**
 * Variable Length Quantity (VLQ) encoding/decoding implementation
 * 
 * In VLQ encoding:
 * - Only the first 7 bits of each byte are used for the actual value
 * - The 8th bit (MSB) is used as a continuation flag:
 *   - If set to 1, it means there are more bytes to follow
 *   - If set to 0, it means this is the last byte in the sequence
 */

/**
 * Encodes an array of integers into VLQ format
 * @param {number[]} numbers - Array of integers to encode
 * @returns {number[]} - Array of bytes in VLQ format
 */
export const encode = (numbers) => {
  const result = [];
  
  for (const num of numbers) {
    // Handle the special case of 0
    if (num === 0) {
      result.push(0);
      continue;
    }
    
    // Convert the number to VLQ format
    const bytes = [];
    let value = num;
    
    // Extract 7 bits at a time, starting from the least significant bits
    while (value > 0) {
      // Take the 7 least significant bits
      bytes.unshift(value & 0x7F);
      // Shift right by 7 bits to process the next 7 bits
      value >>>= 7;
    }
    
    // Set the continuation bit (bit 7) for all bytes except the last one
    for (let i = 0; i < bytes.length - 1; i++) {
      bytes[i] |= 0x80; // Set the MSB to 1
    }
    
    // Add the bytes to the result
    result.push(...bytes);
  }
  
  return result;
};

/**
 * Decodes an array of bytes in VLQ format into an array of integers
 * @param {number[]} bytes - Array of bytes in VLQ format
 * @returns {number[]} - Array of decoded integers
 * @throws {Error} - If the sequence is incomplete
 */
export const decode = (bytes) => {
  const result = [];
  let currentValue = 0;
  let i = 0;
  
  while (i < bytes.length) {
    // Get the current byte
    const byte = bytes[i];
    
    // Extract the 7 least significant bits and add them to the current value
    // Use unsigned right shift to ensure we're treating numbers as unsigned
    currentValue = ((currentValue & 0xFFFFFFF) << 7) | (byte & 0x7F);
    
    // Check if this is the last byte in the sequence (bit 7 is 0)
    if ((byte & 0x80) === 0) {
      // End of sequence, add the value to the result
      result.push(currentValue >>> 0); // Ensure unsigned 32-bit integer
      currentValue = 0;
    } else if (i === bytes.length - 1) {
      // If this is the last byte in the array but the continuation bit is set,
      // then the sequence is incomplete
      throw new Error('Incomplete sequence');
    }
    
    i++;
  }
  
  return result;
};

File: python/variable-length-quantity/variable_length_quantity.py
def encode(numbers):
    """
    Encode a list of integers using Variable Length Quantity (VLQ) encoding.
    
    In VLQ encoding:
    - Only the first 7 bits of each byte are used for data (right-justified)
    - The 8th bit (MSB) is a continuation flag:
      - If MSB is 1, there are more bytes to follow for this number
      - If MSB is 0, this is the last byte for this number
    
    Args:
        numbers: A list of integers to encode
        
    Returns:
        A list of bytes representing the VLQ encoding of the input numbers
    """
    result = []
    
    for number in numbers:
        # Handle the special case of 0
        if number == 0:
            result.append(0)
            continue
            
        # Store the bytes for the current number
        bytes_for_number = []
        
        # Extract 7 bits at a time, starting from the least significant bits
        while number > 0:
            # Get the 7 least significant bits
            byte = number & 0x7F
            number >>= 7
            
            # Add the continuation bit (MSB) for all bytes except the first one we process
            # (which will be the last one in the final sequence)
            if bytes_for_number:
                byte |= 0x80
                
            bytes_for_number.append(byte)
        
        # Reverse the bytes since we processed them from least to most significant
        result.extend(reversed(bytes_for_number))
    
    return result


def decode(bytes_):
    """
    Decode a list of bytes encoded using Variable Length Quantity (VLQ) encoding.
    
    Args:
        bytes_: A list of bytes in VLQ encoding
        
    Returns:
        A list of integers decoded from the input bytes
        
    Raises:
        ValueError: If the sequence is incomplete (i.e., the last byte has MSB=1)
    """
    result = []
    current_number = 0
    
    i = 0
    while i < len(bytes_):
        # Extract the current byte
        byte = bytes_[i]
        
        # Extract the 7 data bits and add them to the current number
        current_number = (current_number << 7) | (byte & 0x7F)
        
        # Check if this is the last byte for the current number
        if not (byte & 0x80):
            # This is the last byte, so add the number to the result and reset
            result.append(current_number)
            current_number = 0
        
        i += 1
    
    # If we've processed all bytes but still have a number being decoded,
    # the sequence is incomplete
    if current_number != 0 or (bytes_ and bytes_[-1] & 0x80):
        raise ValueError("incomplete sequence")
    
    return result
